Allow authentication using RBAC instead of API Keys

adamdougal opened the pull request:
Required  by https://github.com/microsoft/sample-app-aoai-chatGPT/issues/391

pamelafox commented:
Do you also need to add RBAC roles for this to work, like https://github.com/Azure-Samples/azure-search-openai-demo/blob/87d15fc021a2a84c4cbcbec56b2fe0560af05d52/infra/main.bicep#L336 ?

adamdougal commented:
@pamelafox Yes! Looks like I missed that. Some roles are there already but a few are missing. I'll get them added now.

adamdougal commented:
@pamelafox Those have now been added

pamelafox commented on README.md:
```diff
@@ -114,6 +114,17 @@ const getUserInfoList = async () => {
 
+### Authenticate using Azure RBAC
+To authenticate using Azure RBAC instead of the default API Keys follow the steps below:
+1. Ensure role assignments listed on [this page](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/use-your-data?tabs=ai-search#azure-role-based-access-controls-azure-rbac-for-adding-data-sources) have been created 
```
```suggestion
1. Ensure role assignments listed on [this page](https://learn.microsoft.com/azure/ai-services/openai/concepts/use-your-data?tabs=ai-search#azure-role-based-access-controls-azure-rbac-for-adding-data-sources) have been created 
```

pamelafox commented on infra/core/search/search-services.bicep:
```diff
@@ -8,6 +8,7 @@ param sku object = {
 
 param authOptions object = {}
 param semanticSearch string = 'disabled'
+param searchApiKeyEnabled bool = true
```
cc @jongio who likes to be notified about potential changes to infra/core Bicep files (as we try to reuse the same infra/core across azd templates)

pamelafox commented on infra/main.bicep:
```diff
@@ -271,6 +273,46 @@ module searchRoleBackend 'core/security/role.bicep' = {
   }
 }
 
+module searchRoleOpenAi 'core/security/role.bicep' = if (!searchApiKeyEnabled) {
```
I see you only did the system roles, what about local server running? For the other repo with these roles, we also have the equivalent for the user. I don't recall if this repo can be run locally or not.

adamdougal replied:
So I tried locally and it didn't require me to add any roles. I think this is because we're still calling OpenAI with API key. I missed this before, so will make the change and push to this branch/PR along with the associated required role.

jongio replied:
We'll want rbac to work for the local user as well.

jongio commented on README.md:
```diff
@@ -114,6 +114,17 @@ const getUserInfoList = async () => {
 
+### Authenticate using Azure RBAC
+To authenticate using Azure RBAC instead of the default API Keys follow the steps below:
+1. Ensure role assignments listed on [this page](https://learn.microsoft.com/azure/ai-services/openai/concepts/use-your-data?tabs=ai-search#azure-role-based-access-controls-azure-rbac-for-adding-data-sources) have been created 
```
Should mention these should be created by azd provision already

jongio commented on README.md:
````diff
@@ -114,6 +114,17 @@ const getUserInfoList = async () => {
 }
 ```
 
+### Authenticate using Azure RBAC
````
I would rather the flow of the application take care of all the settings and not require the user to take manual action.  

We are experimenting with adding an AZURE_AUTH_TYPE parameter that can get set to rbac or keys and the bicep does the appropriate thing based on that.

It's not finalized, but should be soon.

jongio commented on app.py:
```diff
@@ -215,29 +218,33 @@ def prepare_body_headers_with_data(request):
-                    "strictness": int(AZURE_SEARCH_STRICTNESS)
-                }
-            })
+        cognitiveSearchDataSource = {
```
This should use the search SDK instead of rest

adamdougal replied:
I think that should probably be tackled as a refactor separate to this PR. Happy to make that change if you disagree though.

jongio replied:
It can wait, we are doing in a diff repo, so you could copy that.

sarah-widder replied:
This doesn't relate to the search SDK/API, it's calling the Azure OpenAI chat completions with extensions REST API. I'm working on a refactor to use the OpenAI SDK for this path.

jongio commented on infra/core/ai/cognitiveservices.bicep:
```diff
@@ -41,3 +44,5 @@ output id string = account.id
 output name string = account.name
 output skuName string = account.sku.name
 output key string = account.listKeys().key1
+output identityPrincipalId string = account.identity.principalId
```
This should match the name of the other core files when they provide an id, which is just principalId

jongio commented on infra/core/search/search-services.bicep:
```diff
@@ -17,8 +18,8 @@ resource search 'Microsoft.Search/searchServices@2021-04-01-preview' = {
   properties: {
-    authOptions: authOptions
-    disableLocalAuth: false
+    authOptions: searchApiKeyEnabled ? authOptions : null
```
Param should be something like 'authType' with allowed values of rbac or keys

jongio commented on infra/core/search/search-services.bicep:
```diff
@@ -41,3 +42,4 @@ output endpoint string = 'https://${name}.search.windows.net/'
 output name string = search.name
 output skuName string = sku.name
 output adminKey string = search.listAdminKeys().primaryKey
+output identityPrincipalId string = search.identity.principalId
```
Same comment about name principalId

jongio commented on infra/main.bicep:
```diff
@@ -26,6 +26,7 @@ param searchContentColumns string = 'content'
 param searchFilenameColumn string = 'filepath'
 param searchTitleColumn string = 'title'
 param searchUrlColumn string = 'url'
+param searchApiKeyEnabled bool = true
```
authType

jongio commented on infra/main.bicep:
```diff
@@ -271,6 +273,46 @@ module searchRoleBackend 'core/security/role.bicep' = {
+    principalType: 'ServicePrincipal'
+  }
+}
+
```
In outputs below we shouldn't be outputting keys.  You will get a bicep linter warning.  Best to ask the user to manually set them with in env vars at this time until we have a secure way to store them.

pamelafox replied:
Phew, glad to see the keys are out of this output finally! That's been a TODO (in my head) for a long time.

Another option btw is to provide a script that grabs the keys, which is what I've had to do for another repo which unfortunately requires keys:
https://github.com/Azure-Samples/contoso-chat/pull/33/files#diff-c8719e3915d76118aeee543ed85367b315252d46ef1d793635a82eaef96ce7b4R19

But probably okay not to do that here, since we now have the keyless option.

adamdougal commented:
Heya, thanks for your review comments! I've made some changes. Please note, I've yet to test them out. I'll update on here when I have.

jongio commented on infra/main.parameters.json:
```diff
@@ -43,6 +43,9 @@
       "value": "${AUTH_CLIENT_SECRET}"
+    },
+    "authType": {
+      "value": "${AZURE_AUTH_TYPE}"
```
You'll want to default this to rbac

"value": "${AZURE_AUTH_TYPE=rbac}"

jongio commented on infra/core/search/search-services.bicep:
```diff
@@ -8,7 +8,7 @@ param sku object = {
 param authOptions object = {}
 param semanticSearch string = 'disabled'
-param searchApiKeyEnabled bool = true
+param authType string
```
add rbac default

jongio commented on .env.sample:
```diff
@@ -1,3 +1,4 @@
+AZURE_AUTH_TYPE
```
all the others have an =

adamdougal commented:
I've now tested this deployed to Azure, as well as running locally. I did not have to create any additional role assignments to get this to work locally.

jongio commented:
> I've now tested this deployed to Azure, as well as running locally. I did not have to create any additional role assignments to get this to work locally.

Did you test with both rbac and keys in both local and azure?

adamdougal commented:
> > I've now tested this deployed to Azure, as well as running locally. I did not have to create any additional role assignments to get this to work locally.
> 
> Did you test with both rbac and keys in both local and azure?

Yep, both worked without any additional roles being required.

jongio commented on app.py:
```diff
@@ -788,7 +808,14 @@ def generate_title(conversation_messages):
+        if should_use_keys():
+            openai.api_key = AZURE_OPENAI_KEY
+        else:
+            default_credential = DefaultAzureCredential()
+            token = default_credential.get_token("https://cognitiveservices.azure.com/.default")
+            openai.api_key = token.token
```
Can this be refactored to remove dupe code above?

adamdougal commented:
Heya, I believe this is ready for re-review. Thanks

jongio commented:
Can we move this to azure-samples github org?   Not sure history on why it is in the microsoft org, but most samples go in azure-samples org.  https://github.com/azure-samples

jongio commented on infra/core/host/appservice.bicep:
```diff
@@ -132,6 +132,6 @@ resource applicationInsights 'Microsoft.Insights/components@2020-02-02' existing
 }
 
-output identityPrincipalId string = managedIdentity ? appService.identity.principalId : ''
+output principalId string = managedIdentity ? appService.identity.principalId : ''
```
I stand corrected, the official name is identityPrincipalId

https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/host/appservice.bicep#L121

Please change it back.

jongio commented on infra/core/ai/cognitiveservices.bicep:
```diff
@@ -41,3 +44,5 @@ output id string = account.id
 output name string = account.name
 output skuName string = account.sku.name
 output key string = account.listKeys().key1
+output principalId string = account.identity.principalId
```
Can you change this to identityPrincipalId?  (Sorry)

jongio commented on infra/core/ai/cognitiveservices.bicep:
```diff
@@ -20,6 +20,9 @@ resource account 'Microsoft.CognitiveServices/accounts@2023-05-01' = {
     publicNetworkAccess: publicNetworkAccess
```
Can you do a PR to update this too? https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/ai/cognitiveservices.bicep

adamdougal replied:
What updates are required? Do you just mean to pull from main?

jongio commented on infra/core/search/search-services.bicep:
```diff
@@ -8,6 +8,7 @@ param sku object = {
 
 param authOptions object = {}
 param semanticSearch string = 'disabled'
+param authType string = 'rbac'
```
Can you do a PR to update this? https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/search/search-services.bicep

adamdougal replied:
What updates are required? Do you just mean to pull from main?

jongio commented on infra/main.bicep:
```diff
@@ -265,12 +273,52 @@ module searchRoleBackend 'core/security/role.bicep' = {
+  }
+}
+
+module searchRoleOpenAi 'core/security/role.bicep' = if (authType == 'rbac') {
```
Any new roles you create will also need to be assigned to the user's principalId, see "user roles" above.

adamdougal commented:
Heya @jongio, thanks for the further comments. I have addressed them and rebased from main. 

Unfortunately, I have no information as to why this is not in the azure samples org.

sarah-widder commented:
> Can we move this to azure-samples github org? Not sure history on why it is in the microsoft org, but most samples go in azure-samples org. https://github.com/azure-samples

We have considered moving this to the azure-samples org but haven't taken the action yet. The repo is used in multiple ways (code-first sample, source code for apps deployed from Azure OpenAI studio, storage place for public-facing utility scripts such as data preparation). For now we're keeping it here.

sarah-widder reviewed (approved):
Looks good to me, thank you very much for the contribution.

adamdougal commented:
Heya, can this be merged? Thanks

sarah-widder commented on app.py:
```diff
@@ -41,6 +41,7 @@ def assets(path):
 SEARCH_TOP_K = os.environ.get("SEARCH_TOP_K", 5)
 SEARCH_STRICTNESS = os.environ.get("SEARCH_STRICTNESS", 3)
 SEARCH_ENABLE_IN_DOMAIN = os.environ.get("SEARCH_ENABLE_IN_DOMAIN", "true")
+AZURE_AUTH_TYPE = os.environ.get("AZURE_AUTH_TYPE", "rbac")
```
Can we actually have this default to "keys" and set it to "rbac" from the bicep deployment? The reason to have it default to keys is that for the case of deployment from the Azure OpenAI Studio chat playground, role assignments are not configured by default between the resources, so keys need to be used by default in that scenario.

adamdougal replied:
👍 Good point! I've just pushed that change.

adamdougal commented:
@sarah-widder Is this still required? It looks like similar changes have been made in https://github.com/microsoft/sample-app-aoai-chatGPT/pull/460?

adamdougal commented:
I guess it's just the infra side of things that's still required?

sarah-widder commented:
Hey @adamdougal yes, #460 should have already addressed using RBAC in the app backend, but the infra changes you were working on to automatically assign the right roles during deployment are still needed. If you want to rebase and just keep the infra changes I will give it another review. Thanks very much for your contribution!

adamdougal commented:
I'm going to close this. It looks like the openai version bump has brought significant changes so a fresh PR/branch would probably be a better approach.
